### ISSUE TITLE
Asynchronous routes not being loaded

### ISSUE THREAD

**Issue by Egnus** [EmilTholin/svelte-routing#46]
Hi,
I found out that `svelte-routing` cannot load routes at runtime, is this on purpose or I am doing something wrong?

Here there is an example of 2 routes, exactly the same, but the second one is added after a period of time, this second one will never be loaded:
```html
<script>
    import { pageRoute, errorRoute } from './routes';
    import { config } from './stores';

    let _routes = [];
    let url = '';

    _routes = [..._routes, { path: '/', component: pageRoute }]; // This will work

    config.isLoaded.subscribe(isLoaded => {
        if (isLoaded) {
            _routes = [..._routes, { path: '/', component: pageRoute }]; // after delay, this wont work, even if it is the same component
        }
    });
</script>

<Router {url}>
    <main>
        {#if _routes.length}
            {#each _routes as route}
                <Route path={route.path} component={route.component} />
            {/each}
        {:else}
            <Route path="/" component={errorRoute} />
        {/if}
    </main>
</Router>
```

**Comment by EmilTholin** [comment]
Hi @legionsra!

There seems to be a bug in Svelte. I tried the following code in the [example project](https://github.com/EmilTholin/svelte-routing/tree/master/example), and like you say it doesn't work:

```html
<script>
  import { Router, Route } from "svelte-routing";
  import NavLink from "./components/NavLink.svelte";
  import Home from "./routes/Home.svelte";
  import About from "./routes/About.svelte";
  import Blog from "./routes/Blog.svelte";

  let routes = [
    { path: 'about', component: About },
    { path: 'blog', component: Blog }
  ];

  setTimeout(() => {
    routes = [...routes, { path: '/', component: Home }];
  }, 1000);

  // Used for SSR. A falsy value is ignored by the Router.
  export let url = "";
</script>

<Router url="{url}">
  <nav>
    <NavLink to="/">Home</NavLink>
    <NavLink to="about">About</NavLink>
    <NavLink to="blog">Blog</NavLink>
  </nav>
  <div>
    {#each routes as route}
      <Route path={route.path} component={route.component} />
    {/each}
  </div>
</Router>
```

From [this isolated repro](https://svelte.dev/repl/76f0c759dfe04894935889a4e4950308?version=3.5.1) it seems like reactive declarations are not re-evaluated when a store it depends on is updated asynchronously.

I can open up an issue in Svelte when I get the time, or you can use the minimal repro above and open an issue if you like.

**Comment by legionsra** [comment]
I am writing them in discord now but if I don't manage to find a solution,
it seems that `$: routes =... ` doesn't trigger the revaluation either,
then I will open a ticket.
Of course any relevant info I will update it here.

Thanks.

**Comment by fend25** [comment]
according to docs, `$: routes =... `  won't trigger the revaluation because of `$: ` works only on the top level of <script> block.

Maybe the workarond would be so:

```js
 $routes = writable([
    { path: 'about', component: About },
    { path: 'blog', component: Blog }
  ]);

  setTimeout(() => {
    routes.set([...routes, { path: '/', component: Home }]);
  }, 1000);
```

**Comment by EmilTholin** [comment]
Hi @fend25! That won't work either, sadly. The issue is that the [underlying reactive statement in the `Router` component](https://github.com/EmilTholin/svelte-routing/blob/17dec56959728bddb91583b3868604a9eeb51500/src/Router.svelte#L98-L105) isn't re-evaluated when the routes store is updated asynchronously.

**Comment by Egnus** [comment]
perhaps there is a way we can change this on the code?

**Comment by Egnus** [comment]
any idea with this issue?

**Comment by Egnus** [comment]
I debugged the code and, considering that `$:` is running at the root scope, I personally considered not to rely on Svelte to re-evaluate a function based on changes of local variables.

There might be a bug for this in svelte, but in this case I think the best approach for a store that depends on another 2 stores is to use `derived` instead which was the idea it was designed for.
So instead of:
```javascript
$: {
    const bestMatch = pick($routes, $location.pathname);
    activeRoute.set(bestMatch);
  }
```
I think it should be better, shorter and more in control like this:
```javascript
   const activeRoute =
      typeof window === 'undefined' 
        ? writable(null)
        : derived([routes, location], ([routes, location]) => pick(routes, location.pathname));
```

If you agree please feel free to merge this PR: https://github.com/EmilTholin/svelte-routing/pull/48

**Comment by EmilTholin** [comment]
There is an issue in the Svelte issue tracker for this: https://github.com/sveltejs/svelte/issues/3284

I have commented with a minimal reproduction of what gives rise to the bug in this library. When that is fixed this will work as intended.

**Comment by sqpdln** [comment]
curious if anyone found a workaround for this?

**Comment by EmilTholin** [comment]
This has been fixes upstream in `svelte@3.16.5`, so if you upgrade to that asynchronous routes should work as expected.
